Thread safe eagerloading

adambedford opened the pull request:
This replaces the `require` statements with thread-safe eager autoloading provided by ActiveSupport.

Commit:
Eager autoload constants for thread-safety

Commit:
Require specific ActiveSupport Autoload library

Commit:
Merge branch 'master' into thread-safe-eagerloading

ridiculous commented:
From my understanding of `ActiveSupport::Autoload` this will only be thread safe if we call:

``` ruby
NameDrop.eager_load!
NameDrop::Resources.eager_load!
```

at the end of `name_drop.rb` which equates to just requiring them ... So, what's the benefit of this?

adambedford commented:
The rails app can register the namespaces it wants to eager load as follows:

```
config.eager_load_namespaces << [NameDrop, NameDrop::Resources]
```

But you bring up a good point. What we actually should do here is override `NameDrop.eager_load!` to eager load the other modules, and then add a railtie to add the `NameDrop` module to the `eager_load_namespaces`.

Watch this space...

prpetten commented on lib/name_drop/configuration.rb:
```diff
@@ -15,18 +15,4 @@ def initialize
     end
   end
-
-  class << self
```
I know that these methods belong to the `NameDrop` module, and not the `NameDrop::Configuration` class.  However, I liked logically grouping them into the `lib/name_drop/configuration.rb` file that contains all the other `configuration`, as I thought it was easier to understand with all of it in one place than adding them into the `lib/name_drop.rb` file.

ridiculous replied:
I'm on board with that. Only thing I'd request is that we at least have this have in the main `name_drop.rb`:

``` ruby
module NameDrop
end
```

adambedford replied:
I'm of the opinion that the file and the module should be tightly coupled, so you know where to look for everything related to that module.

Can roll this back though if there are strong opposing views?

ridiculous replied:
Up to @prpetten, I don't have a strong opinion on this

prpetten replied:
I see that side of the argument; no need to roll it back.

ridiculous commented:
@adambedford cool, that sounds good to me 👍 I think this PR should include the Railtie, so it brings it full circle

Commit:
Add railtie to add NameDrop module to `eager_load_namespaces` in a Rails environment

Automatically eager load `NameDrop::Resources` when `NameDrop` module is eager loaded

Commit:
Merge branch 'master' into thread-safe-eagerloading

adambedford commented on lib/name_drop/configuration.rb:
```diff
@@ -15,18 +15,4 @@ def initialize
     end
   end
-
-  class << self
```
👍 then this should be good to go!

Sent from my iPhone

> On Sep 12, 2016, at 5:38 PM, Paul Pettengill notifications@github.com wrote:
> 
> In lib/name_drop/configuration.rb:
> 
> > @@ -15,18 +15,4 @@ def initialize
> >        @access_token = 'NotARealAccessToken'
> >      end
> > 
> > ##    end
> > -  class << self
> >   I see that side of the argument; no need to roll it back.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
